### PR TITLE
[Wallet] Performance optimizations, memory usage revamp

### PR DESCRIPTION
**Pull request by bvbfan**
It does not contain braking changes, it redesigns containers usage to prevent memory overhead. I will plan to reindex over this branch for testing. @uzyn will notify you when all looks good to be tested on microservice node to verify it will reduce memory usage. @ShengguangXiao can also reindex over this branch on main and testnet, observe memusage as well.

### CODE REVIEW

**Comment by bvbfan** [comment]
I bit more aggressive refactor of dead wallet code. Strip out unused code, minimize CWalletTx size.

**Comment by uzyn** [comment]
@bvbfan the community found something: https://www.reddit.com/r/defiblockchain/comments/l765tb/fix_for_ram_bug_and_corrupt_blockchain/

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -100,13 +99,9 @@ static std::set<std::string> g_unloading_wallet_set;
 // Custom deleter for shared_ptr<CWallet>.
 static void ReleaseWallet(CWallet* wallet)
 {
-    // Unregister and delete the wallet right after BlockUntilSyncedToCurrentChain
-    // so that it's in sync with the current chainstate.
     const std::string name = wallet->GetName();
     wallet->WalletLogPrintf("Releasing wallet\n");
-    wallet->BlockUntilSyncedToCurrentChain();
     wallet->Flush();
-    wallet->m_chain_notifications_handler.reset();
```
Comment (ShengguangXiao) [inline src/wallet/wallet.cpp:109]: Why these two lines are removed?

File: src/wallet/wallet.h [hunk 2]
```diff
@@ -679,6 +680,28 @@ struct CoinSelectionParams
...
+            bm::tag<ByHash>, bm::const_mem_fun<CWalletTx, const uint256&, &CWalletTx::GetHash>
+        >,
+        bm::ordered_non_unique<
+            bm::tag<ByOrder>, bm::member<CWalletTx, int64_t, &CWalletTx::nOrderPos>
+        >,
+        bm::ordered_non_unique<
+            bm::tag<ByTime>, bm::member<CWalletTx, unsigned int, &CWalletTx::nTimeReceived>
+        >
+    >
+> mapWalletType;
```
Comment (ShengguangXiao) [inline src/wallet/wallet.h:703]: The type change to `multiIndexWalletType` will be more readable.

File: src/wallet/walletdb.h [hunk 3]
```diff
@@ -185,7 +185,7 @@ class WalletBatch
             return false;
         }
         m_database.IncrementUpdateCounter();
-        if (m_database.nUpdateCounter % 1000 == 0) {
+        if (m_database.nUpdateCounter % 100 == 0) {
```
Comment (ShengguangXiao) [inline src/wallet/walletdb.h:188]: Can we define 100 as a const member variable?

File: src/wallet/wallet.cpp (hunk 1 above)
Comment (bvbfan) [reply src/wallet/wallet.cpp:109]: It makes wallet unstable. They are in `UnloadWallet`

File: src/wallet/wallet.h (hunk 2 above)
Comment (bvbfan) [reply src/wallet/wallet.h:703]: Underlying type multi_index isn't needed, about me.

File: src/wallet/walletdb.h (hunk 3 above)
Comment (bvbfan) [reply src/wallet/walletdb.h:188]: Will revert it. Let stay 1000.

**Comment by bvbfan** [comment]
Round 2 of cleaning wallet dead code. Now it uses mineCached as well.

File: src/wallet/walletdb.h (hunk 3 above)
Comment (ShengguangXiao) [reply src/wallet/walletdb.h:188]: Now the value still 100

**Comment by ShengguangXiao** [comment]
I test the change, reindex blocks and wallets works fine, because my wallet is small, so didn't see much ram difference as before.

**Comment by monstrobishi** [comment]
@bvbfan there is a problem in travis in `rpc_fundrawtransaction.py              | ✖ Failed  | 59 s`

**Comment by bvbfan** [comment]
It can be tested on systems with many addresses and/or txs, wallet should be more responsive, full sync uses same memory on my end.

File: src/sync.h [hunk 4]
```diff
@@ -318,4 +319,22 @@ struct SCOPED_LOCKABLE LockAssertion
...
+    std::atomic_bool& lock;
+public:
+    CLockFreeGuard(std::atomic_bool& lock) : lock(lock)
+    {
+        bool desired = false;
+        while (!lock.compare_exchange_weak(desired, true,
+                                           std::memory_order_release,
+                                           std::memory_order_relaxed))
+            desired = false;
+    }
```
Comment (ShengguangXiao) [inline src/sync.h:333]: This loop will consume a lot of CPU resources.

**Comment by monstrobishi** [comment]
@bvbfan still have a failing test here: wallet_basic.py", line 367

File: src/sync.h (hunk 4 above)
Comment (bvbfan) [reply src/sync.h]: Force pushed a fix.

**Comment by bvbfan** [comment]
Tested on full sync, it should be not a wip anymore.

**Pull request merged**